src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
+        return temp;
+    }
+
+    private void linkBefore(T value, Node<T> node) {
+        Node<T> newNode = new Node<>(node.prev, value, node);
+        if (node.prev == null) {
+            first = newNode;
+        } else {
+            node.prev.next = newNode;
+        }
+        node.prev = newNode;
+        size++;
+    }
+
+    private void checkElementIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("There is no element with this index");
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("This index is invalid");
+        }
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] arrayElements = list.toArray();
+        Node<T> prev = last;
+        for (Object object : arrayElements) {
+            T element = (T) object;
+            Node<T> newNode = new Node<>(prev, element, null);
+            if (prev == null) {
+                first = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            prev = newNode;
+            size++;
+        }
+        last = prev;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> temp = node(index);
+        T oldItem = temp.item;
+        temp.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> temp = node(index);
+        Node<T> next = temp.next;
+        Node<T> prev = temp.prev;
+        final T element = temp.item;
+        if (prev == null) {
+            first = temp.next;
+        } else {
+            prev.next = next;
+            temp.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            temp.next = null;
+        }
+        temp.item = null;
+        size--;
+        return element;
     }
 
     @Override
     public boolean remove(T object) {
+        int i = 0;
+        if (object == null) {
+            for (Node<T> temp = first; temp != null; temp = temp.next, i++) {
+                if (temp.item == null) {
+                    remove(i);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> temp = first; temp != null; temp = temp.next, i++) {
```

Comment:
complicated for loop: hard to read such solution. 
try to use a while loop.
you don't need an `int i` because `Node` contains the index.